src/libsystemd/sd-event/sd-event.c

```diff
@@ -3795,29 +3795,44 @@ _public_ int sd_event_wait(sd_event *e, uint64_t timeout) {
                 return 1;
         }
 
-        event_queue_max = MAX(e->n_sources, 1u);
-        if (!GREEDY_REALLOC(e->event_queue, e->event_queue_allocated, event_queue_max))
+        n_event_queue = MAX(e->n_sources, 1u);
+        if (!GREEDY_REALLOC(e->event_queue, e->event_queue_allocated, n_event_queue))
                 return -ENOMEM;
 
         /* If we still have inotify data buffered, then query the other fds, but don't wait on it */
         if (e->inotify_data_buffered)
-                timeout = 0;
+                msec = 0;
+        else
+                msec = timeout == (uint64_t) -1 ? -1 : (int) DIV_ROUND_UP(timeout, USEC_PER_MSEC);
 
-        m = epoll_wait(e->epoll_fd, e->event_queue, event_queue_max,
-                       timeout == (uint64_t) -1 ? -1 : (int) DIV_ROUND_UP(timeout, USEC_PER_MSEC));
-        if (m < 0) {
-                if (errno == EINTR) {
-                        e->state = SD_EVENT_PENDING;
-                        return 1;
+        for (m = 0;;) {
+                r = epoll_wait(e->epoll_fd, e->event_queue + m, e->event_queue_allocated - m, msec);
+                if (r < 0) {
+                        if (errno == EINTR) {
+                                e->state = SD_EVENT_PENDING;
+                                return 1;
+                        }
+
+                        r = -errno;
+                        goto finish;
                 }
 
-                r = -errno;
-                goto finish;
+                m += r;
+                if (m < e->event_queue_allocated)
+                        break;
+
+                if (e->event_queue_allocated >= n_event_queue * 10)
+                        break;
+
+                if (!GREEDY_REALLOC(e->event_queue, e->event_queue_allocated, e->event_queue_allocated + n_event_queue))
+                        return -ENOMEM;
+
+                msec = 0;
         }
```

Comment:
i don't think this can work. i.e. if we don#t actually act on any of the fds between the epoll calls they'll remain queued, and when we query again they'll still be returned to us. that's at least how i understood this. thus if you do this like this the returned queue will contain the same entries over and over again, each time only with a few more added in (not sure if added to the end or to the middle though).